server/src/main/java/org/elasticsearch/cluster/action/index/MappingUpdatedAction.java

```diff
@@ -57,34 +60,50 @@ private void setDynamicMappingUpdateTimeout(TimeValue dynamicMappingUpdateTimeou
         this.dynamicMappingUpdateTimeout = dynamicMappingUpdateTimeout;
     }
 
-
     public void setClient(Client client) {
         this.client = client.admin().indices();
     }
 
-    private PutMappingRequestBuilder updateMappingRequest(Index index, String type, Mapping mappingUpdate, final TimeValue timeout) {
-        if (type.equals(MapperService.DEFAULT_MAPPING)) {
-            throw new IllegalArgumentException("_default_ mapping should not be updated");
-        }
-        return client.preparePutMapping().setConcreteIndex(index).setType(type).setSource(mappingUpdate.toString(), XContentType.JSON)
-                .setMasterNodeTimeout(timeout).setTimeout(TimeValue.ZERO);
-    }
-
-    /**
-     * Same as {@link #updateMappingOnMaster(Index, String, Mapping, TimeValue)}
-     * using the default timeout.
-     */
-    public void updateMappingOnMaster(Index index, String type, Mapping mappingUpdate) {
-        updateMappingOnMaster(index, type, mappingUpdate, dynamicMappingUpdateTimeout);
-    }
-
     /**
      * Update mappings on the master node, waiting for the change to be committed,
      * but not for the mapping update to be applied on all nodes. The timeout specified by
      * {@code timeout} is the master node timeout ({@link MasterNodeRequest#masterNodeTimeout()}),
      * potentially waiting for a master node to be available.
      */
-    public void updateMappingOnMaster(Index index, String type, Mapping mappingUpdate, TimeValue masterNodeTimeout) {
-        updateMappingRequest(index, type, mappingUpdate, masterNodeTimeout).get();
+    public void updateMappingOnMaster(Index index, String type, Mapping mappingUpdate, ActionListener<Void> listener) {
+        if (type.equals(MapperService.DEFAULT_MAPPING)) {
+            throw new IllegalArgumentException("_default_ mapping should not be updated");
+        }
+        client.preparePutMapping().setConcreteIndex(index).setType(type).setSource(mappingUpdate.toString(), XContentType.JSON)
+            .setMasterNodeTimeout(dynamicMappingUpdateTimeout).setTimeout(TimeValue.ZERO)
+            .execute(new ActionListener<AcknowledgedResponse>() {
+                @Override
+                public void onResponse(AcknowledgedResponse acknowledgedResponse) {
+                    listener.onResponse(null);
+                }
+
+                @Override
+                public void onFailure(Exception e) {
+                    listener.onFailure(unwrapException(e));
+                }
+            });
+    }
+
+    private static Exception unwrapException(Exception e) {
```

Comment:
This is a little strange admittedly, but I needed part of the exception unwrapping logic from `org.elasticsearch.common.util.concurrent.FutureUtils#get(java.util.concurrent.Future<T>, long, java.util.concurrent.TimeUnit)` to get the same exception rendering as we currently have with the blocking call so I copied it. Maybe it would make sense to provide a version of `execute` that does this unwrapping automatically to make the blocking and non-blocking calls behave the same way on failures eventually?